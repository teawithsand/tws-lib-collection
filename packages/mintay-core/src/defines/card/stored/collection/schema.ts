import { z } from "zod"

const StoredSdelkaCollectionDataV1Schema = z.object({
	globalId: z.string(),
	title: z.string(),
	description: z.string(),
	createdAtTimestamp: z.number(),
	lastUpdatedAtTimestamp: z.number(),
})

export const StoredSdelkaCollectionDataSchema = z.discriminatedUnion(
	"version",
	[
		z.object({
			version: z.literal(1),
			data: StoredSdelkaCollectionDataV1Schema,
		}),
	],
)

export type StoredSdelkaCollectionData = z.infer<
	typeof StoredSdelkaCollectionDataSchema
>
