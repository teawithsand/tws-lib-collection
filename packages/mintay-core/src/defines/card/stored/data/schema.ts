import { z } from "zod"

const StoredSdelkaCardDataV1Schema = z.object({
	globalId: z.string(),
	content: z.string(),
	createdAtTimestamp: z.number(),
	lastUpdatedAtTimestamp: z.number(),
})

export const StoredSdelkaCardDataSchema = z.discriminatedUnion("version", [
	z.object({
		version: z.literal(1),
		data: StoredSdelkaCardDataV1Schema,
	}),
])

export type StoredSdelkaCardData = z.infer<typeof StoredSdelkaCardDataSchema>
