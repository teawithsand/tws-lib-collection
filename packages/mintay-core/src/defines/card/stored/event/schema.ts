import { z } from "zod"
import { StoredSdelkaCardEventV1Schema } from "./schemaV1"

export const StoredSdelkaCardEventSchema = z.discriminatedUnion("version", [
	z.object({
		version: z.literal(1),
		data: StoredSdelkaCardEventV1Schema,
	}),
])

export type StoredSdelkaCardEvent = z.infer<typeof StoredSdelkaCardEventSchema>
