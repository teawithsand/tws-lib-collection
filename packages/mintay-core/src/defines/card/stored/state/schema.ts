import { z } from "zod"
import { StoredSdelkaCardStateV1Schema } from "./schemaV1"

export const StoredSdelkaCardStateSchema = z.discriminatedUnion("version", [
	z.object({
		version: z.literal(1),
		data: StoredSdelkaCardStateV1Schema,
	}),
])

export type StoredSdelkaCardState = z.infer<typeof StoredSdelkaCardStateSchema>
