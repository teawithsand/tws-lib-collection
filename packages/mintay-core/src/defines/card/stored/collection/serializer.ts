import { SdelkaCollectionData } from "../../sdelkaCollectionData"
import { StoredSdelkaCollectionData } from "./schema"

const neverGuard = (value: never): never => {
	throw new Error(`Unsupported version: ${value}`)
}

export class StoredSdelkaCollectionDataSerializer {
	private constructor() {}

	public static readonly serialize = (
		data: SdelkaCollectionData,
	): StoredSdelkaCollectionData => {
		return {
			version: 1,
			data: {
				globalId: data.globalId,
				title: data.title,
				description: data.description,
				createdAtTimestamp: data.createdAtTimestamp,
				lastUpdatedAtTimestamp: data.lastUpdatedAtTimestamp,
			},
		}
	}

	public static readonly deserialize = (
		storedData: StoredSdelkaCollectionData,
	): SdelkaCollectionData => {
		if (storedData.version === 1) {
			const d = storedData.data
			return {
				globalId: d.globalId,
				title: d.title,
				description: d.description,
				createdAtTimestamp: d.createdAtTimestamp,
				lastUpdatedAtTimestamp: d.lastUpdatedAtTimestamp,
			}
		}
		return neverGuard(storedData.version)
	}
}
