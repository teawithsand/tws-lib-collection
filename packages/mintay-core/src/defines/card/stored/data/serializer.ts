import { SdelkaCardData } from "../../sdelkaCardData"
import { StoredSdelkaCardData } from "./schema"

const neverGuard = (value: never): never => {
	throw new Error(`Unsupported version: ${value}`)
}

export class StoredSdelkaCardDataSerializer {
	private constructor() {}

	public static readonly serialize = (
		data: SdelkaCardData,
	): StoredSdelkaCardData => {
		return {
			version: 1,
			data: {
				globalId: data.globalId,
				content: data.content,
				createdAtTimestamp: data.createdAtTimestamp,
				lastUpdatedAtTimestamp: data.lastUpdatedAtTimestamp,
			},
		}
	}

	public static readonly deserialize = (
		storedData: StoredSdelkaCardData,
	): SdelkaCardData => {
		if (storedData.version === 1) {
			const d = storedData.data
			return {
				globalId: d.globalId,
				content: d.content,
				createdAtTimestamp: d.createdAtTimestamp,
				lastUpdatedAtTimestamp: d.lastUpdatedAtTimestamp,
			}
		}
		return neverGuard(storedData.version)
	}
}
