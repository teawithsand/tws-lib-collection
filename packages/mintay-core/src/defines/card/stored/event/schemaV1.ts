import { z } from "zod"

export enum StoredSdelkaCardEventTypeV1 {
	ANSWER = 0,
}

export enum StoredSdelkaAnswerV1 {
	AGAIN = 1,
	HARD = 2,
	GOOD = 3,
	EASY = 4,
}

export const StoredSdelkaCardEventV1Schema = z.object({
	type: z.nativeEnum(StoredSdelkaCardEventTypeV1),
	answer: z.nativeEnum(StoredSdelkaAnswerV1),
	timestamp: z.number(),
})

export type StoredSdelkaCardEventV1 = z.infer<
	typeof StoredSdelkaCardEventV1Schema
>
