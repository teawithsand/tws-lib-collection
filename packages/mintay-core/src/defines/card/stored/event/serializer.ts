import { SdelkaAnswer } from "../../answer"
import { SdelkaCardEvent, SdelkaCardEventType } from "../../sdelkaCardEvent"
import { StoredSdelkaCardEvent } from "./schema"
import { StoredSdelkaAnswerV1, StoredSdelkaCardEventTypeV1 } from "./schemaV1"

export class SdelkaCardEventSerializer {
	public static readonly toStoredAnswer = (
		answer: SdelkaAnswer,
	): StoredSdelkaAnswerV1 => {
		switch (answer) {
			case SdelkaAnswer.AGAIN:
				return StoredSdelkaAnswerV1.AGAIN
			case SdelkaAnswer.HARD:
				return StoredSdelkaAnswerV1.HARD
			case SdelkaAnswer.GOOD:
				return StoredSdelkaAnswerV1.GOOD
			case SdelkaAnswer.EASY:
				return StoredSdelkaAnswerV1.EASY
		}
		const _exhaustiveCheck: never = answer
		return _exhaustiveCheck
	}

	public static readonly fromStoredAnswer = (
		answer: StoredSdelkaAnswerV1,
	): SdelkaAnswer => {
		switch (answer) {
			case StoredSdelkaAnswerV1.AGAIN:
				return SdelkaAnswer.AGAIN
			case StoredSdelkaAnswerV1.HARD:
				return SdelkaAnswer.HARD
			case StoredSdelkaAnswerV1.GOOD:
				return SdelkaAnswer.GOOD
			case StoredSdelkaAnswerV1.EASY:
				return SdelkaAnswer.EASY
		}
		const _exhaustiveCheck: never = answer
		return _exhaustiveCheck
	}

	public static readonly serialize = (
		event: SdelkaCardEvent,
	): StoredSdelkaCardEvent => {
		switch (event.type) {
			case SdelkaCardEventType.ANSWER:
				return {
					version: 1,
					data: {
						type: StoredSdelkaCardEventTypeV1.ANSWER,
						answer: SdelkaCardEventSerializer.toStoredAnswer(
							event.answer,
						),
						timestamp: event.timestamp,
					},
				}
		}
		const _exhaustiveCheck: never = event.type
		return _exhaustiveCheck
	}

	public static readonly deserialize = (
		storedEvent: StoredSdelkaCardEvent,
	): SdelkaCardEvent => {
		if (storedEvent.version === 1) {
			const data = storedEvent.data
			switch (data.type) {
				case StoredSdelkaCardEventTypeV1.ANSWER:
					return {
						type: SdelkaCardEventType.ANSWER,
						answer: SdelkaCardEventSerializer.fromStoredAnswer(
							data.answer,
						),
						timestamp: data.timestamp,
					}
			}
			const _exhaustiveCheck: never = data.type
			return _exhaustiveCheck
		}
		const _exhaustiveCheckVersion: never = storedEvent.version
		return _exhaustiveCheckVersion
	}
}
