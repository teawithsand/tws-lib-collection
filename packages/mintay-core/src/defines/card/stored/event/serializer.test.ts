import { describe, expect, test } from "vitest"
import { SdelkaAnswer } from "../../answer"
import { SdelkaCardEventType } from "../../sdelkaCardEvent"
import { StoredSdelkaCardEvent } from "./schema"
import { StoredSdelkaAnswerV1, StoredSdelkaCardEventTypeV1 } from "./schemaV1"
import { SdelkaCardEventSerializer } from "./serializer"

describe("SdelkaCardEventSerializer", () => {
	const timestamp = 1234567890

	const answersToTest = [
		SdelkaAnswer.AGAIN,
		SdelkaAnswer.HARD,
		SdelkaAnswer.GOOD,
		SdelkaAnswer.EASY,
	]

	const eventTypesToTest = [
		SdelkaCardEventType.ANSWER,
		// Add other event types here if they exist
	]

	eventTypesToTest.forEach((eventType) => {
		answersToTest.forEach((answer) => {
			test(`serialize and deserialize with eventType ${eventType} and answer ${answer}`, () => {
				const event = {
					type: eventType,
					answer: answer,
					timestamp,
				}
				const serialized = SdelkaCardEventSerializer.serialize(event)
				expect(serialized.data.type).toBe(
					StoredSdelkaCardEventTypeV1.ANSWER,
				)
				expect(serialized.data.answer).toBe(answer)
				const deserialized =
					SdelkaCardEventSerializer.deserialize(serialized)
				expect(deserialized.type).toBe(eventType)
				expect(deserialized.answer).toBe(answer)
				expect(deserialized.timestamp).toBe(timestamp)
				expect(deserialized).toEqual(event)
			})
		})
	})

	describe("serialize", () => {
		test("should serialize ANSWER event correctly", () => {
			const event = {
				type: SdelkaCardEventType.ANSWER,
				answer: SdelkaAnswer.GOOD,
				timestamp,
			}
			const stored = SdelkaCardEventSerializer.serialize(event)
			expect(stored.version).toBe(1)
			expect(stored.data.type).toBe(StoredSdelkaCardEventTypeV1.ANSWER)
			expect(stored.data.answer).toBe(StoredSdelkaAnswerV1.GOOD)
			expect(stored.data.timestamp).toBe(timestamp)
		})

		test("serialize and then deserialize should not change the event", () => {
			const event = {
				type: SdelkaCardEventType.ANSWER,
				answer: SdelkaAnswer.GOOD,
				timestamp,
			}
			const stored = SdelkaCardEventSerializer.serialize(event)
			const deserialized = SdelkaCardEventSerializer.deserialize(stored)
			expect(deserialized).toEqual(event)
		})

		test("serialize snapshot", () => {
			const event = {
				type: SdelkaCardEventType.ANSWER,
				answer: SdelkaAnswer.GOOD,
				timestamp,
			}
			const stored = SdelkaCardEventSerializer.serialize(event)
			expect(stored).toMatchSnapshot()
		})
	})

	describe("deserialize", () => {
		test("should deserialize ANSWER event correctly", () => {
			const storedEvent: StoredSdelkaCardEvent = {
				version: 1,
				data: {
					type: StoredSdelkaCardEventTypeV1.ANSWER,
					answer: StoredSdelkaAnswerV1.GOOD,
					timestamp,
				},
			}
			const event = SdelkaCardEventSerializer.deserialize(storedEvent)
			expect(event.type).toBe(SdelkaCardEventType.ANSWER)
			expect(event.answer).toBe(SdelkaAnswer.GOOD)
			expect(event.timestamp).toBe(timestamp)
		})

		test("deserialize and then serialize should not change the stored event", () => {
			const storedEvent: StoredSdelkaCardEvent = {
				version: 1,
				data: {
					type: StoredSdelkaCardEventTypeV1.ANSWER,
					answer: StoredSdelkaAnswerV1.GOOD,
					timestamp,
				},
			}
			const event = SdelkaCardEventSerializer.deserialize(storedEvent)
			const reStored = SdelkaCardEventSerializer.serialize(event)
			expect(reStored).toEqual(storedEvent)
		})

		test("deserialize snapshot", () => {
			const storedEvent: StoredSdelkaCardEvent = {
				version: 1,
				data: {
					type: StoredSdelkaCardEventTypeV1.ANSWER,
					answer: StoredSdelkaAnswerV1.GOOD,
					timestamp,
				},
			}
			const event = SdelkaCardEventSerializer.deserialize(storedEvent)
			expect(event).toMatchSnapshot()
		})
	})
})
