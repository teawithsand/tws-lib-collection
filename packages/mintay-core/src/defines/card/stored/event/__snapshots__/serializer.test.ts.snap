// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SdelkaCardEventSerializer > deserialize > deserialize snapshot 1`] = `
{
  "answer": 3,
  "timestamp": 1234567890,
  "type": 0,
}
`;

exports[`SdelkaCardEventSerializer > serialize > serialize snapshot 1`] = `
{
  "data": {
    "answer": 3,
    "timestamp": 1234567890,
    "type": 0,
  },
  "version": 1,
}
`;
